### PR TITLE
Added 'Native methods Over Util' article

### PR DESCRIPTION
**Pull request by Berkmann18**
This PR adds the "Prefer native JS methods over user-land utils like Lodash" article (including the inclusion in the README).

re #302

### CODE REVIEW

**Comment by i0natan** [comment]
@sagirk @BrunoScheufler @js-kyle 

Since it's the first perf BP in collaboration with a contributor, I'd like to all of your thoughts on how to smoothen this process. Specifically, what should we emphasize in our writing guidelines to minimize back and forth?

@BrunoScheufler What about our linter? 😠 😄

File: README.md [hunk 1]
```diff
@@ -955,6 +955,14 @@ All statements above will return false if used with `===`
 
 ## Our contributors are working on this section. [Would you like to join?](https://github.com/i0natan/nodebestpractices/issues/256)
 
+## ![✔] 7.1. Prefer native JS methods over user-land utils like Lodash
+
+**TL;DR:** It's often more penalising to use utility libraries like `lodash` and `underscore` over using native methods as it leads to uneeded dependencies with less performance boost.
```
Comment (BrunoScheufler) [inline README.md]:
```suggestion
 **TL;DR:** It's often more penalising to use utility libraries like `lodash` and `underscore` over using native methods as it leads to unneeded dependencies with less performance boost.
```

File: sections/performance/nativeoverutil.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+```
+You can find a bigger list of benchmarks [here](https://github.com/Berkmann18/NativeVsUtils/blob/master/index.txt) or alternatively [run this](https://github.com/Berkmann18/NativeVsUtils/blob/master/index.js) which would show the same but with colours.
+
+### "You don't (may not) need Lodash/Underscore"
+
+From the [repo on this matter which focuses on Lodash and Underscore](https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore).
+
+ > Lodash and Underscore are great modern JavaScript utility libraries, and they are widely used by Front-end developers. However, when you are targeting modern browsers, you may find out that there are many methods which are already supported natively thanks to ECMAScript5 [ES5] and ECMAScript2015 [ES6]. If you want your project to require fewer dependencies, and you know your target browser clearly, then you may not need Lodash/Underscore.
+
+There's also an [ESLint plugin](https://www.npmjs.com/package/eslint-plugin-you-dont-need-lodash-underscore) which detects where you're using either libraries but don't need to.
```
Comment (BrunoScheufler) [inline sections/performance/nativeoverutil.md]:
```suggestion
 There's also an [ESLint plugin](https://www.npmjs.com/package/eslint-plugin-you-dont-need-lodash-underscore) which detects where you're using libraries but don't need to.
```

File: sections/performance/nativeoverutil.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+Here's what ESLint would output when using the YDNLU plugin.
+```bash
+/home/maxie/GDrive/Other/Sandbox/YDNLU/lodashLove.js
+  5:13  warning  Consider using the native Array.prototype.map()       you-dont-need-lodash-underscore/map
+  7:5   warning  Consider using the native Array.prototype.includes()  you-dont-need-lodash-underscore/includes
+
+✖ 2 problems (0 errors, 2 warnings)
+```
+
+Of course, the example above doesn't seem realistic considering to what actual codebases would have but you get the idea.
```
Comment (BrunoScheufler) [inline sections/performance/nativeoverutil.md]:
```suggestion
 Of course, the example above doesn't seem realistic considering what actual codebases would have but you get the idea.
```

**Comment by Berkmann18** [comment]
@BrunoScheufler I applied the changes you requested.

**Comment by BrunoScheufler** [comment]
@Berkmann18 thanks for your contribution!

@i0natan I think this is good to go, what do you think? If you agree, you can simply merge it 😄

**Comment by Berkmann18** [comment]
@BrunoScheufler You're welcome :smile:.

File: README.md [hunk 4]
```diff
@@ -955,6 +955,14 @@ All statements above will return false if used with `===`
 
 ## Our contributors are working on this section. [Would you like to join?](https://github.com/i0natan/nodebestpractices/issues/256)
 
+## ![✔] 7.1. Prefer native JS methods over user-land utils like Lodash
+
+ **TL;DR:** It's often more penalising to use utility libraries like `lodash` and `underscore` over native methods as it leads to unneeded dependencies and slower performance.
+
+**Otherwise:** You'll have to maintain (slightly) bigger projects where you could have simply used what was **already** available or dealt with a few more lines in exchange of a few more files.
```
[the remark this answers is not in the document]
Comment (Berkmann18) [reply README.md]: Sure, I've done that.
Is this:
> You'll have to maintain less performant projects where you could have simply used what was **already** available or dealt with a few more lines in exchange of a few more files.
Fine?

File: sections/performance/nativeoverutil.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+```
+You can find a bigger list of benchmarks [here](https://github.com/Berkmann18/NativeVsUtils/blob/master/index.txt) or alternatively [run this](https://github.com/Berkmann18/NativeVsUtils/blob/master/index.js) which would show the same but with colours.
+
+### "You don't (may not) need Lodash/Underscore"
+
+From the [repo on this matter which focuses on Lodash and Underscore](https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore).
+
+ > Lodash and Underscore are great modern JavaScript utility libraries, and they are widely used by Front-end developers. However, when you are targeting modern browsers, you may find out that there are many methods which are already supported natively thanks to ECMAScript5 [ES5] and ECMAScript2015 [ES6]. If you want your project to require fewer dependencies, and you know your target browser clearly, then you may not need Lodash/Underscore.
+
+ There's also an [ESLint plugin](https://www.npmjs.com/package/eslint-plugin-you-dont-need-lodash-underscore) which detects where you're using libraries but don't need to.
```
[the remark this answers is not in the document]
Comment (Berkmann18) [reply sections/performance/nativeoverutil.md]: Good point, I've added that title.

File: sections/performance/nativeoverutil.md [hunk 6]
```diff
@@ -0,0 +1,90 @@
+# Prefer native JS methods over user-land utils like Lodash
+
+<br/><br/>
+
+### One Paragraph Explainer
+
+Sometimes, using native methods is better than requiring `lodash` or `underscore` because the resulting program will have more dependencies and thus use more space and also not guarantee the best performance all-round.
```
[the remark this answers is not in the document]
Comment (Berkmann18) [reply sections/performance/nativeoverutil.md]: I've taken that into account and simplified the paragraph which emphasise the performance loss.

File: sections/performance/nativeoverutil.md [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+console.log(_.map(arr, x => `d${x}`));
+
+if (_.includes(arr, 0)) console.log('0 found');
+
+console.log('compacted:', _.compact(arr));
+```
+
+Here's what ESLint would output when using the YDNLU plugin.
+```bash
+/home/maxie/GDrive/Other/Sandbox/YDNLU/lodashLove.js
```
[the remark this answers is not in the document]
Comment (Berkmann18) [reply sections/performance/nativeoverutil.md]: Good point.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (i0natan) [reply README.md]: @Berkmann18 Is this the opening sentence?

File: README.md [hunk 8]
```diff
@@ -955,6 +955,15 @@ All statements above will return false if used with `===`
 
 ## Our contributors are working on this section. [Would you like to join?](https://github.com/i0natan/nodebestpractices/issues/256)
 
+## ![✔] 7.1. Prefer native JS methods over user-land utils like Lodash
+
+ **TL;DR:** It's often more penalising to use utility libraries like `lodash` and `underscore` over native methods as it leads to unneeded dependencies and slower performance.
+ Bar in mind that with the introduction of the new V8 engine alongside the new ES standards, native methods were improved in such a way that it's now about 50% more performant than utility libraries.
```
Comment (i0natan) [inline README.md]: @Berkmann18 @BrunoScheufler Bar is a TYPE or some English slang?

File: sections/performance/nativeoverutil.md [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+### One Paragraph Explainer
+
+Sometimes, using native methods is better than requiring `lodash` or `underscore` because the resulting program will have more dependencies and thus use more space and also not guarantee the best performance all-round.
+
+<!-- comp here: https://gist.github.com/Berkmann18/3a99f308d58535ab0719ac8fc3c3b8bb-->
+
+<br/><br/>
+
+### Code Example – Benchmark test on `_.concat`/`Array.concat`
```
[the remark this answers is not in the document]
Comment (i0natan) [reply sections/performance/nativeoverutil.md]: @Berkmann18 At your own convenience, try thinking how to shorten and sharpen this example. I know how hard it's, found myself more than a few times trying to cut in a half a complex code sample

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (Berkmann18) [reply README.md]: It's the sentence for the "**Otherwise**" section.

File: sections/performance/nativeoverutil.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (Berkmann18) [reply sections/performance/nativeoverutil.md]: I found a way to shorten that example without hiding too much (say, if someone were to copy-paste the code without missing necessary bits).

File: README.md [hunk 10]
```diff
@@ -955,6 +955,15 @@ All statements above will return false if used with `===`
 
 ## Our contributors are working on this section. [Would you like to join?](https://github.com/i0natan/nodebestpractices/issues/256)
 
+## ![✔] 7.1. Prefer native JS methods over user-land utils like Lodash
+
+ **TL;DR:** It's often more penalising to use utility libraries like `lodash` and `underscore` over native methods as it leads to unneeded dependencies and slower performance.
+ Bear in mind that with the introduction of the new V8 engine alongside the new ES standards, native methods were improved in such a way that it's now about 50% more performant than utility libraries.
```
Comment (i0natan) [inline README.md]: Can we link to this evidence?

File: sections/performance/nativeoverutil.md [hunk 11]
```diff
@@ -0,0 +1,76 @@
+# Prefer native JS methods over user-land utils like Lodash
+
+
+<br/><br/>
+
+### One Paragraph Explainer
+
+Sometimes, using native methods is better than requiring `lodash` or `underscore` because it **will not lead in a performance boost** and use more space than necessary.
+The performance using native methods result in an overall ~50% gain which includes the following methods:
+ - `Array.concat`
```
Comment (i0natan) [inline sections/performance/nativeoverutil.md]: Try keeping the format like other bullets: no bold, no lists - just plain text organized with line breaks

File: sections/performance/nativeoverutil.md [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+
+<!-- comp here: https://gist.github.com/Berkmann18/3a99f308d58535ab0719ac8fc3c3b8bb-->
+
+<br/><br/>
+
+### Example: benchmark comparison - Lodash vs V8 (Native)
+The graph below shows the mean of the benchmarks for a variety of Lodash methods, this shows that Lodash methods take on average 146.23% more time to complete the same tasks as V8 methods.
+
+![meanDiag](../../assets/images/sampleMeanDiag.png)
```
Comment (i0natan) [inline sections/performance/nativeoverutil.md]: Let's include a link to the article?

File: sections/performance/nativeoverutil.md [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+  .run({ 'async': true });
+```
+
+Which returns this:
+
+![output](../../assets/images/concat-benchmark.png)
+
+You can find a bigger list of benchmarks [here](https://github.com/Berkmann18/NativeVsUtils/blob/master/index.txt) or alternatively [run this](https://github.com/Berkmann18/NativeVsUtils/blob/master/index.js) which would show the same but with colours.
+
+### "You don't (may not) need Lodash/Underscore"
```
Comment (i0natan) [inline sections/performance/nativeoverutil.md]: Add "Blog Quote:" before the title, [see other bullets for example](https://github.com/i0natan/nodebestpractices/blob/master/sections/projectstructre/breakintcomponents.md)

File: sections/performance/nativeoverutil.md [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+
+From the [repo on this matter which focuses on Lodash and Underscore](https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore).
+
+ > Lodash and Underscore are great modern JavaScript utility libraries, and they are widely used by Front-end developers. However, when you are targeting modern browsers, you may find out that there are many methods which are already supported natively thanks to ECMAScript5 [ES5] and ECMAScript2015 [ES6]. If you want your project to require fewer dependencies, and you know your target browser clearly, then you may not need Lodash/Underscore.
+
+ There's also an [ESLint plugin](https://www.npmjs.com/package/eslint-plugin-you-dont-need-lodash-underscore) which detects where you're using libraries but don't need to.
+
+Here's an example of that plugin in use:
+Consider a file called _lodashLove.js_ shown below
```
Comment (i0natan) [inline sections/performance/nativeoverutil.md]: This example shouldn't be more than 2,3 **simpler** lines, just call 2 methods and add a comment explaining that eslint plugin will flag an error

File: sections/performance/nativeoverutil.md [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+
+You can find a bigger list of benchmarks [here](https://github.com/Berkmann18/NativeVsUtils/blob/master/index.txt) or alternatively [run this](https://github.com/Berkmann18/NativeVsUtils/blob/master/index.js) which would show the same but with colours.
+
+### "You don't (may not) need Lodash/Underscore"
+
+From the [repo on this matter which focuses on Lodash and Underscore](https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore).
+
+ > Lodash and Underscore are great modern JavaScript utility libraries, and they are widely used by Front-end developers. However, when you are targeting modern browsers, you may find out that there are many methods which are already supported natively thanks to ECMAScript5 [ES5] and ECMAScript2015 [ES6]. If you want your project to require fewer dependencies, and you know your target browser clearly, then you may not need Lodash/Underscore.
+
+ There's also an [ESLint plugin](https://www.npmjs.com/package/eslint-plugin-you-dont-need-lodash-underscore) which detects where you're using libraries but don't need to.
```
Comment (i0natan) [inline sections/performance/nativeoverutil.md]: Create a new section, new paragraph, new Example: Linting for non-native methods usage. Separate this from the quote @Berkmann18

File: README.md (hunk 10 above)
Comment (Berkmann18) [reply README.md]: Sure, I'll add a link to the spreadsheet containing the data and will try to make the result easier to spot.

Comment (i0natan) [reply README.md]: @Berkmann18 Is this some online blog post or some private Excel file? we have the graph which you show below, maybe link to it?

File: sections/performance/nativeoverutil.md (hunk 12 above)
Comment (Berkmann18) [reply sections/performance/nativeoverutil.md]: I've added a link to the relevant spreadsheet if that's fine?

File: sections/performance/nativeoverutil.md (hunk 15 above)
Comment (Berkmann18) [reply sections/performance/nativeoverutil.md]: Okay, I've revisited in that part and the example below that part to make it coherent.

**Comment by i0natan** [comment]
@Berkmann18 Perfect. A star will be assigned by the end of January #318 🏅 

I will also Tweet this new BP today.

Hope to see you here again soon :]

**Comment by Berkmann18** [comment]
Woo! I'm glad I contributed and look forward to contributing more.
